Add support for reading (most) process-related CSV files

alexdewar opened the pull request:
# Description

This is a rough draft of code to read in most of the process-related CSV files. I generated the data structures with a Python script to guess the column types, then edited the results. Most of my effort was spent on integrating the data structures together, which isn't something we've done elsewhere in the codebase so far. The result is that we have a `read_processes()` function, which returns a `Vec<Process>`. Each `Process` struct also contains vectors representing its associated availabilities, flows, PACs, parameters and regions. I omitted code for reading the flow share and investment constraints files following @ahawkes' suggestion (the example files are currently also empty), but I'll open separate issues to solve these.

I also noticed that the first column in `processes.csv` is called `process_id`, whereas it should be `id`, so I've fixed that.

Questions:

* Is this way of representing the processes sensible, in terms of how we'll be using this data? While one could extract e.g. the availabilities for all processes from this data structure, it's not ideal, so if we will be accessing these vectors outside of the context of the process to which they relate, then we might want to consider alternatives.
* The `flow_cost` column in `process_flows.csv` is currently empty. I've currently assumed that that means they should all be treated as zero, but if the empty field means something else, then we should probably represent it as `None`. If we *do* want it to be zero, I'd suggest being explicit and filling the column with zeroes.

Not implemented yet:

* Unit tests
* Check for duplicate `id` values in `processes.csv`

Closes #64.

## Type of change

- [ ] Bug fix (non-breaking change to fix an issue)
- [x] New feature (non-breaking change to add functionality)
- [ ] Refactoring (non-breaking, non-functional change to improve maintainability)
- [ ] Optimization (non-breaking change to speed up the code)
- [ ] Breaking change (whatever its nature)
- [ ] Documentation (improve or add documentation)

## Key checklist

- [x] All tests pass: `$ cargo test`
- [ ] The documentation builds and looks OK: `$ cargo doc`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

ahawkes commented:
Thanks looks good. Regarding questions:

1) At the moment I can't think of instances where we will access this data out of scope of the process. But that may arise in future. It does still feel like RDB approach might be cleaner, where all this is accessed via (well written) SQL, and we carry around a pointer to the DB. I'm not sure the overhead of accessing these structs - it's likely we'll be needing the data in them a lot.

2) I think I prefer an approach where fields can be blank where they are optional, and the default is specified in the documentation. At the moment we have noted some (perhaps not all!) of these cases in the data schema https://lucid.app/lucidchart/6e05ea99-97d3-45f0-b1cf-2a300eff21e7/edit?viewport_loc=-1472%2C-330%2C3923%2C2319%2C0_0&invitationId=inv_ad88b9f5-799d-4bf6-9e60-1559aa1688fc

Also - I see that where time slice availability isn't provided you are assuming a "None" value. I think the default should be the overall annual availability (i.e. sum of output over a year divided by theoretical maximum output over a year).

ahawkes reviewed (approved):
Looks good.

alexdewar commented:
@ahawkes Thanks for the review!

> Thanks looks good. Regarding questions:
> 
> 1. At the moment I can't think of instances where we will access this data out of scope of the process. But that may arise in future. It does still feel like RDB approach might be cleaner, where all this is accessed via (well written) SQL, and we carry around a pointer to the DB. I'm not sure the overhead of accessing these structs - it's likely we'll be needing the data in them a lot.

I'd like to talk longer about this on Monday when we meet, but I've looked into trying to integrate an in-memory database here and essentially it's not going to be a good fit for the project. It's definitely worth spending time considering what data structures will serve us best (I'm including an RDB as a data structure here), but to make that decision, we really need to know how this data will be accessed. Getting the data into memory is easy enough, but it needs to be in a form that's easy to work with, for the specific tasks we'll be performing. (The performance considerations are also rather different when going from Python to Rust too: basically in Python you try to avoid manually writing loops as much as possible for scientific computing in favour of using data frames, numpy arrays etc., because the performance is horrible for a large number of iterations, but this isn't a consideration for Rust in the same way.) I still think this "vector of structs" approach is the right one for now, though it *may* be that `polars` proves useful depending on the complexity of the queries we're making. As I said, I think this needs a longer discussion, but it is worth thinking about how we will be accessing the data (i.e. what the higher-level interface for the input layer will be), as well as how many time slices, agents, regions etc. we're likely to encounter in users' models, and then it will become more obvious which data structures will serve that purpose best.

> 2. I think I prefer an approach where fields can be blank where they are optional, and the default is specified in the documentation. At the moment we have noted some (perhaps not all!) of these cases in the data schema https://lucid.app/lucidchart/6e05ea99-97d3-45f0-b1cf-2a300eff21e7/edit?viewport_loc=-1472%2C-330%2C3923%2C2319%2C0_0&invitationId=inv_ad88b9f5-799d-4bf6-9e60-1559aa1688fc

Thanks for this. I'll add these default values in before we merge.

> 
> Also - I see that where time slice availability isn't provided you are assuming a "None" value. I think the default should be the overall annual availability (i.e. sum of output over a year divided by theoretical maximum output over a year).

This is just a placeholder for now, but in the longer term I'd like to have this field be an enum or reference to a `TimeSlice` struct, rather than just treating it as an `Option<String>` and I'll make sure to use that as the default value when I do!

alexdewar commented:
Actually, on second thoughts, @dalonsoa, dw about reviewing this for now. I'll finish it up properly then mark this as ready for review.

alexdewar commented:
Changes since last review:

- Make some fields optional with default values
- Add more input checks
- Use enum types for fields which can only have set values (e.g. `flow_type`)

I'm afraid it's become a little on the long side. Sorry @dalonsoa :disappointed:

dalonsoa reviewed (approved):
I have a few questions and a suggestion, but otherwise it looks good. Once you get around the weird syntax, it is pretty cool.

dalonsoa commented:
BUT, there are not that many unit test, just a big integration test. If you're not implementing those now, open an issue to do so in the future.

alexdewar commented on examples/simple/commodities.csv:
All the various IDs (e.g. for agents too etc.) seem to be in uppercase, but there's no reason they have to be from the perspective of the code (insofar as we're actually using any of them).

Not sure why `type` and `timeslice_level` are in uppercase, but that's how things are described in @ahawkes and @tsmbland's documents.

Personally, I think these should be lowercase -- we have other "enum-type" fields in other files that are -- but either way we should probably be consistent. We want to read in these values in a case-insensitive way (that's what I've done with the enum types in this PR), so it won't matter from the code's perspective, but let's be consistent in our examples/documentation.

Is everyone happy with this?

alexdewar replied:
I've taken the liberty of opening an issue for it: #120

alexdewar commented on examples/simple/commodity_costs.csv:
`time_slice` refers to a specific time slice, i.e. time of year + time of day, whereas `time_slice_level` indicates whether it's the time of year or time of day that is relevant for this row. Does that make sense?

alexdewar commented on src/process.rs:
```diff
@@ -0,0 +1,291 @@
+    pub discount_rate: f64,
+    pub cap2act: f64,
+}
+define_id_getter! {ProcessParameter}
```
It uses the macro I defined above, which in this case will expand to:

```rust
        impl HasProcessID for ProcessParameter {
            fn get_process_id(&self) -> &str {
                &self.process_id
            }
        }
```

Basically we need to define a trait for all of these structs in order to be able to pass them into `read_csv_grouped_by_id` which wants to be able to get their process IDs (even though the field is called `process_id` for all of them!).

alexdewar commented on src/process.rs:
```diff
@@ -0,0 +1,291 @@
+    let mut parameters = read_csv_grouped_by_id_with_filter(
+        process_parameters_file_path,
+        &process_ids,
+        |file_path, param: ProcessParameterRaw| param.into_parameter(file_path, &year_range),
```
Exactly. It's not obvious though, so I'll add a description to the doc comment.

alexdewar commented on src/process.rs:
```diff
@@ -0,0 +1,291 @@
+            Process {
+                id: desc.0,
+                description: desc.1,
+                availabilities: availabilities.remove(id.as_str()).unwrap_or_default(),
```
The key type for `description` is `String` whereas for the others it's a `&str`. Usually you can just pass a `String` in these cases and it'll get converted, but for some reason that didn't work so I had to do it explicitly :shrug:
